invoice/invoice.py

```diff
@@ -61,41 +61,35 @@ def get_origin_line_key(self, name):
         return ''
 
 
-class InvoiceReport(HTMLReport):
+class InvoiceReport(HTMLReportMixin, metaclass=PoolMeta):
     __name__ = 'account.invoice'
 
     @classmethod
-    def __setup__(cls):
-        super(InvoiceReport, cls).__setup__()
-        cls.__rpc__['execute'] = RPC(False)
+    def _execute_html_report(cls, records, data, action, side_margin=2,
+            extra_vertical_margin=30):
 
-    @classmethod
-    def execute(cls, ids, data):
         pool = Pool()
         Invoice = pool.get('account.invoice')
 
-        action, _ = cls.get_action(data)
-
-        if len(ids) == 1:
-            # Re-instantiate because records are TranslateModel
-            invoice, = Invoice.browse(ids)
-            if invoice.invoice_report_cache:
-                return (
-                    invoice.invoice_report_format,
-                    bytes(invoice.invoice_report_cache),
-                    cls.get_direct_print(action),
-                    cls.get_name(action))
-
-        result = super(InvoiceReport, cls).execute(ids, data)
-
-        if (len(ids) == 1 and invoice.state in {'posted', 'paid'}
-                and invoice.type == 'out'):
-            with Transaction().set_context(_check_access=False):
-                invoice, = Invoice.browse([invoice.id])
-                format_, data = result[0], result[1]
-                invoice.invoice_report_format = format_
-                invoice.invoice_report_cache = \
-                    Invoice.invoice_report_cache.cast(data)
-                invoice.save()
-
-        return result
+        # Re-instantiate because records are DualRecord
+        invoice, = Invoice.browse([r.raw.id for r in records])
```

Comment:
There could be more than one record, so this would crash.

I think that just like in the original code we should only use the cache if we're requested a single record.

Reply:
See d6b90203d4ddd25718465c2b111599928f26b6b7

In case invoice report is not single and user do print multiples records, we need to return report cache or report rendered.